modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SSource.java

```diff
@@ -312,7 +322,14 @@ void write(MethodWriter writer, Globals globals) {
         }
 
         if (!methodEscape) {
-            writer.visitInsn(Opcodes.ACONST_NULL);
+            if (scriptInterface.getExecuteMethodReturnType().equals(Definition.DEF_TYPE)) {
+                // TODO this should fail too but it can't to keep compatibility with Elasticsearch scripts
+                writer.visitInsn(Opcodes.ACONST_NULL);
+            } else if (false == scriptInterface.getExecuteMethodReturnType().equals(Definition.VOID_TYPE)) {
+                assert false : "this should have failed during the analyze phase";
+                throw new IllegalArgumentException("Expected all paths to [return] but not all did or end in an expression to return but "
+                        + "some path ends is missing a return and ends in a statement.");
+            }
```

Comment:
After looking at this code, I think I'd prefer default values for primitives to keep consistency with other types, and possibly have one for all non-primitive types instead of just def that will return null.  Once we're closer to having custom contexts we can probably re-evaluate this and remove the default auto-return since I do appreciate a high-level of strictness.